File: tools/hollow-knight-patches/src/skins/core/SkinResourceModels.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace DualSouls.Skins.HollowKnight.Core
{
    // Trusted verifier facts; this value does not verify files or canonical paths.
    public sealed record VerifiedResourceSource(string ProofScope, string CanonicalSource, long EncodedLength,
        string Sha256, long Width, long Height, string Format = "RGBA32", ResourceUsage Usage = ResourceUsage.GAME);
    public enum ResourceUsage { GAME, PREVIEW }
    public enum ResourceOwnership { OWNED, BORROWED }
    public enum ResourceReferenceKind { TARGET, MATERIAL, ROLLBACK }
    public enum ResourceAllocationPhase { RESERVED, READY }
    public sealed record ResourceReference(long Id, string PlanId, long AllocationId, ResourceReferenceKind Kind);
    public sealed record ResourceDecodeOperation(long OperationId, long AllocationId, string PlanId);
    public sealed record ResourceScratch(ResourceDecodeOperation Operation, long Bytes, bool DecodeCompleted = false);
    public sealed class ResourceAllocation
    {
        public long Id { get; }
        public ResourceOwnership Ownership { get; }
        public long Width { get; }
        public long Height { get; }
        public IReadOnlyList<VerifiedResourceSource> Sources { get; }
        public IReadOnlyList<string> PlanIds { get; }
        public string BorrowedKey { get; }
        public ResourceAllocationPhase Phase { get; }
        public ResourceAllocation(long id, ResourceOwnership ownership, long width, long height,
            IEnumerable<VerifiedResourceSource> sources = null, IEnumerable<string> planIds = null,
            string borrowedKey = null, ResourceAllocationPhase phase = ResourceAllocationPhase.RESERVED)
        {
            Id = id; Ownership = ownership; Width = width; Height = height; BorrowedKey = borrowedKey; Phase = phase;
            Sources = Array.AsReadOnly((sources ?? Array.Empty<VerifiedResourceSource>()).ToArray());
            PlanIds = Array.AsReadOnly((planIds ?? Array.Empty<string>()).ToArray());
        }
    }
    // Admission ledger only: NOT preparation, commit, retirement or apply readiness proof.
    // Plan membership retains its charge at zero references. Lifetime/disposal is separate.
    public sealed class ResourceState
    {
        public IReadOnlyList<ResourceAllocation> Allocations { get; }
        public IReadOnlyList<ResourceReference> References { get; }
        public ResourceScratch Scratch { get; }
        public long NextId { get; }
        public ResourceState(IEnumerable<ResourceAllocation> allocations = null, IEnumerable<ResourceReference> references = null,
            ResourceScratch scratch = null, long nextId = 1)
        {
            Allocations = Array.AsReadOnly((allocations ?? Array.Empty<ResourceAllocation>()).ToArray());
            References = Array.AsReadOnly((references ?? Array.Empty<ResourceReference>()).ToArray());
            Scratch = scratch; NextId = nextId;
        }
    }
    public sealed class ResourceTotals
    {
        public long ResidentBytes { get; }
        public long ProcessBytes { get; }
        public int OwnedAllocations { get; }
        public IReadOnlyDictionary<string, long> PlanBytes { get; }
        public IReadOnlyDictionary<string, int> PlanAllocations { get; }
        public ResourceTotals(long residentBytes, long processBytes, int ownedAllocations,
            IDictionary<string, long> planBytes, IDictionary<string, int> planAllocations)
        {
            ResidentBytes = residentBytes; ProcessBytes = processBytes; OwnedAllocations = ownedAllocations;
            PlanBytes = new ReadOnlyDictionary<string, long>(new Dictionary<string, long>(planBytes));
            PlanAllocations = new ReadOnlyDictionary<string, int>(new Dictionary<string, int>(planAllocations));
        }
    }
    public abstract record ResourceEvent
    {
        private ResourceEvent() { }
        public sealed record Acquire(string PlanId, VerifiedResourceSource Source, ResourceReferenceKind Kind) : ResourceEvent;
        public sealed record AcquireBorrowed(string PlanId, string BorrowedKey, long Width, long Height, ResourceReferenceKind Kind) : ResourceEvent;
        public sealed record ReleaseReference(ResourceReference Reference) : ResourceEvent;
        // Trusted acknowledgement of the exact decode operation, not simulated proof.
        public sealed record DecodeCompleted(ResourceDecodeOperation Operation) : ResourceEvent;
        // Acknowledges release of ALL readable CPU upload/scratch for this operation.
        public sealed record ScratchReleased(ResourceDecodeOperation Operation) : ResourceEvent;
    }
    public abstract record ResourceCommand
    {
        private ResourceCommand() { }
        // Fixed RGBA32, no mipmaps. Value command only; does not decode.
        public sealed record Decode(ResourceDecodeOperation Operation, VerifiedResourceSource Source) : ResourceCommand;
        public sealed record ReferenceGranted(ResourceReference Reference) : ResourceCommand;
        public sealed record ReleaseUploadScratch(ResourceDecodeOperation Operation) : ResourceCommand;
    }
    public sealed class ResourceDecision
    {
        public ResourceState State { get; }
        public bool Accepted { get; }
        public string Diagnosis { get; }
        public IReadOnlyList<ResourceCommand> Commands { get; }
        public ResourceDecision(ResourceState state, bool accepted, string diagnosis, params ResourceCommand[] commands)
        {
            State = state; Accepted = accepted; Diagnosis = diagnosis;
            Commands = Array.AsReadOnly((ResourceCommand[])commands.Clone());
        }
    }
}

File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinResourceModels.kt
package dev.silksong.launcher.skins.core

import java.util.Collections

/** Facts supplied by a trusted verifier; this value does not verify files or canonical paths. */
data class VerifiedResourceSource(
    val proofScope: String, val canonicalSource: String, val encodedLength: Long,
    val sha256: String, val width: Long, val height: Long,
    val format: String = "RGBA32", val usage: ResourceUsage = ResourceUsage.GAME,
)
enum class ResourceUsage { GAME, PREVIEW }
enum class ResourceOwnership { OWNED, BORROWED }
enum class ResourceReferenceKind { TARGET, MATERIAL, ROLLBACK }
enum class ResourceAllocationPhase { RESERVED, READY }
data class ResourceReference(val id: Long, val planId: String, val allocationId: Long, val kind: ResourceReferenceKind)
data class ResourceDecodeOperation(val operationId: Long, val allocationId: Long, val planId: String)
data class ResourceScratch(val operation: ResourceDecodeOperation, val bytes: Long, val decodeCompleted: Boolean = false)

class ResourceAllocation(
    val id: Long, val ownership: ResourceOwnership, val width: Long, val height: Long,
    sources: Collection<VerifiedResourceSource> = emptyList(), planIds: Collection<String> = emptyList(),
    val borrowedKey: String? = null, val phase: ResourceAllocationPhase = ResourceAllocationPhase.RESERVED,
) {
    val sources: List<VerifiedResourceSource> = Collections.unmodifiableList(sources.toList())
    val planIds: List<String> = Collections.unmodifiableList(planIds.toList())
}

/** Admission ledger only, NOT proof of preparation, commit, retirement or apply readiness.
 * Plan membership retains its budget charge even when its last reference is released.
 * Zero-reference allocations remain process-charged: lifetime/disposal is a separate slice.
 */
class ResourceState(
    allocations: Collection<ResourceAllocation> = emptyList(), references: Collection<ResourceReference> = emptyList(),
    val scratch: ResourceScratch? = null, val nextId: Long = 1,
) {
    val allocations: List<ResourceAllocation> = Collections.unmodifiableList(allocations.toList())
    val references: List<ResourceReference> = Collections.unmodifiableList(references.toList())
}
class ResourceTotals(
    val residentBytes: Long, val processBytes: Long, val ownedAllocations: Int,
    planBytes: Map<String, Long>, planAllocations: Map<String, Int>,
) {
    val planBytes: Map<String, Long> = Collections.unmodifiableMap(planBytes.toMap())
    val planAllocations: Map<String, Int> = Collections.unmodifiableMap(planAllocations.toMap())
}
sealed interface ResourceEvent {
    data class Acquire(val planId: String, val source: VerifiedResourceSource, val kind: ResourceReferenceKind) : ResourceEvent
    data class AcquireBorrowed(val planId: String, val borrowedKey: String, val width: Long, val height: Long, val kind: ResourceReferenceKind) : ResourceEvent
    data class ReleaseReference(val reference: ResourceReference) : ResourceEvent
    /** Trusted adapter acknowledgement of this exact allocation operation, not simulated proof. */
    data class DecodeCompleted(val operation: ResourceDecodeOperation) : ResourceEvent
    /** Acknowledges release of ALL readable CPU upload/scratch for this operation. */
    data class ScratchReleased(val operation: ResourceDecodeOperation) : ResourceEvent
}
sealed interface ResourceCommand {
    /** Fixed RGBA32, no mipmaps. Value command only; does not perform decoding. */
    data class Decode(val operation: ResourceDecodeOperation, val source: VerifiedResourceSource) : ResourceCommand
    data class ReferenceGranted(val reference: ResourceReference) : ResourceCommand
    data class ReleaseUploadScratch(val operation: ResourceDecodeOperation) : ResourceCommand
}
class ResourceDecision(val state: ResourceState, val accepted: Boolean, val diagnosis: String, commands: Collection<ResourceCommand> = emptyList()) {
    val commands: List<ResourceCommand> = Collections.unmodifiableList(commands.toList())
}
